Generify lister-gen

skitt opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

This adds a generic implementation of a lister, and uses it to replace the template code in generated listers. The corresponding templates are no longer used and are removed.

Listers are reduced to their interfaces (non-namespaced and namespaced if appropriate), their specific structs, and their constructors. All method implementations are provided by the generic implementation. The dedicated interface is preserved so that each lister can have its own set of methods (e.g. the method returning the namespaced lister if appropriate), and the dedicated struct is preserved to allow expansions to be defined where necessary.

The external interface is unchanged and doesn't expose generics.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

The core of this change is https://github.com/kubernetes/kubernetes/pull/121574/files#diff-8b2fc7ac28a79479e5ffb0b072f0b5d1b308ff4cb29daa3ae0657e23d200ac46 and https://github.com/kubernetes/kubernetes/pull/121574/files#diff-08b8fe190a6fe29b3cb7708a75aa9e7256459c4f7d941854d76eae79862ff56f — the rest of the change is generated code.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

skitt committed:
Generify lister-gen

This adds a generic implementation of a lister, and uses it to replace
the template code in generated listers. The corresponding templates
are no longer used and are removed.

Listers are reduced to their interfaces (non-namespaced and namespaced
if appropriate), their specific structs, and their constructors. All
method implementations are provided by the generic implementation. The
dedicated interface is preserved so that each lister can have its own
set of methods (e.g. the method returning the namespaced lister if
appropriate), and the dedicated struct is preserved to allow
expansions to be defined where necessary.

The external interface is unchanged and doesn't expose generics.

Signed-off-by: Stephen Kitt <skitt@redhat.com>

skitt commented:
@thockin if you have the time I’m interested in your thoughts on this and #121439 — I understand you like generics but are also wary of their overuse.

thockin commented on staging/src/k8s.io/client-go/generic/listers.go:
```diff
@@ -0,0 +1,87 @@
+	resource schema.GroupResource
+}
+
+// Namespace wraps a lister and namespace for a given type.
```
comment and code don't match (type name)

thockin reviewed (commented):
I don't find this abusive AT ALL.  I think this is exactly what generics are good for, and anything that reduces the LOC count of generated code is a win.  

I'd love to see us explore a "v2" client which really leans into generics (even at the user-facing API) and uses less code generation.  These codegen tools are crusty and hard to maintain and even harder to change (hence why my go-workspaces work has stalled).

thockin commented:
@liggitt @jpbetz @deads2k

skitt commented:
/ok-to-test

seans3 commented:
/triage accepted

skitt commented:
I’ve updated the PR to address the linting issues, and also to leave `client-go/generic` free for a possible future generic client (alongside `dynamic` etc.). The helpers are moved into `listers`, and renamed to reflect their nature (for `struct`s) and reduce stuttering (for constructors).

thockin commented:
If this were broken into commits with "real work" and "regenerated code" it would help review.

Who is the right approver for this?  I love the idea, but I am not deeply familiar with this area

thockin commented on staging/src/k8s.io/client-go/listers/generic_helpers.go:
```diff
@@ -0,0 +1,87 @@
+	"k8s.io/client-go/tools/cache"
+)
+
+// ResourceIndexer wraps an indexer and resource for a given type.
```
Some comments on how this is used might be great - e.g. is it for end-users?  Or is it used to implement user-facing APis, but is not reall meant to be user-facing itself?

thockin commented on staging/src/k8s.io/client-go/listers/generic_helpers.go:
```diff
@@ -0,0 +1,87 @@
+)
+
+// ResourceIndexer wraps an indexer and resource for a given type.
+type ResourceIndexer[T runtime.Object] struct {
```
API topic.  The way this is currently used is like:

```
listers.ResourceIndexer[*v1.FooBar]
```

In particular, the `*` is part of the type argument.  Could/should we simplfy that into:

```
listers.ResourceIndexer[v1.FooBar]
```

...which uses the actual type name, not the pointer.  Then signatures like:

```
func (l ResourceIndexer[T]) List(selector labels.Selector) (ret []*T, err error) {
  // ...
}

func (l ResourceIndexer[T]) Get(name string) (*T, error) {
  // ...
}
```

I personally find this more natural, but I have been known to get a bit too familiar with generics.

skitt replied:
That doesn’t work here because `T` is based on `runtime.Object`, and the latter’s methods all have pointer receivers, so `T` has to be a pointer type.

thockin commented on staging/src/k8s.io/code-generator/cmd/lister-gen/generators/lister.go:
```diff
@@ -286,45 +283,21 @@ type $.type|public$Lister interface {
 var typeListerStruct = `
 // $.type|private$Lister implements the $.type|public$Lister interface.
 type $.type|private$Lister struct {
```
Can we document why we choose to embed:

```
type fooLister struct {
    listers.ResourceIndexer[foo]
}
```

vs. alias:

```
type fooLister = listers.ResourceIndexer[foo]
```

?  If there are tradeoffs, can we put them in a comment here?

skitt replied:
I mentioned this in the commit message and PR description, but I’ve added a comment above both structs.

thockin commented on staging/src/k8s.io/code-generator/cmd/lister-gen/generators/lister.go:
```diff
@@ -346,31 +319,6 @@ var namespaceListerStruct = `
-		return nil, errors.NewNotFound($.Resource|raw$("$.type|lowercaseSingular$"), name)
-	}
-	return obj.(*$.type|raw$), nil
+	listers.NamespacedResourceIndexer[*$.type|raw$]
```
same comment as above re: embed vs alias?

thockin commented on staging/src/k8s.io/kube-aggregator/pkg/client/listers/apiregistration/v1/apiservice.go:
```diff
@@ -39,30 +39,10 @@ type APIServiceLister interface {
 
 // aPIServiceLister implements the APIServiceLister interface.
 type aPIServiceLister struct {
```
LOL at aPIService.  Can we file a bug to fix the namer to recognize "API" -> "api"

skitt committed:
Regenerate all listers

Signed-off-by: Stephen Kitt <skitt@redhat.com>

skitt commented:
> If this were broken into commits with "real work" and "regenerated code" it would help review.
> 
> Who is the right approver for this? I love the idea, but I am not deeply familiar with this area

This particular generator hasn’t been touched in a while, perhaps @tiran or @sttts or @alvaroaleman...

thockin commented:
Where does this PR stand?  I am a strong +1.

skitt commented:
> Where does this PR stand? I am a strong +1.

It’s ready to go as far as I’m concerned, it just needs reviewing by potential approvers.

skitt commented:
@thockin I’ve rebased this on your gengo v2 work.

thockin commented:
I love it.  Let's pinky-promise to merge this ASAP once the window opens.

/lgtm
/approve

pacoxu commented:
/hold
for https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/batch/pull-kubernetes-verify/1780819179771793408

https://prow.k8s.io/?repo=kubernetes%2Fkubernetes&type=batch
/test pull-kubernetes-verify

skitt commented:
Thanks @pacoxu, this does indeed need a last-minute codegen re-run before merging to make sure all the generated code is up-to-date.

@thockin this means the PR needs a new lgtm :wink:

pacoxu commented:
/unhold

skitt commented:
/retest-required

thockin commented on staging/src/k8s.io/client-go/listers/generic_helpers.go:
```diff
@@ -0,0 +1,91 @@
+func (l NamespacedResourceIndexer[T]) Get(name string) (T, error) {
+	obj, exists, err := l.indexer.GetByKey(l.namespace + "/" + name)
+	if err != nil {
+		return *new(T), err
```
why don't we return nil?

thockin commented on staging/src/k8s.io/client-go/listers/generic_helpers.go:
```diff
@@ -0,0 +1,91 @@
+		return *new(T), err
+	}
+	if !exists {
+		return *new(T), errors.NewNotFound(l.resource, name)
```
same

thockin commented on staging/src/k8s.io/client-go/listers/generic_helpers.go:
```diff
@@ -0,0 +1,91 @@
+
+// NamespacedResourceIndexer wraps an indexer, resource, and namespace for a given type.
+// This is intended for use by listers (generated by lister-gen) only.
+type NamespacedResourceIndexer[T runtime.Object] struct {
```
Idea:

Why not just have 1 type:

```
ResourceIndexer[T runtime.Object] struct {
	indexer  cache.Indexer
	resource schema.GroupResource
	namespace string // empty for non-namespaced types
}
```

and then just test `l.namespace` for empty.  One code path, one type. Two
constructors functions (`New` and `NewNamespaced`).

skitt replied:
I thought I had tried that and run into issues, but apparently not; I’ve updated the PR to use a single type.

thockin reviewed (commented):
I'm picking nits now.

skitt commented on staging/src/k8s.io/client-go/listers/generic_helpers.go:
```diff
@@ -0,0 +1,72 @@
+
+// List lists all resources in the indexer matching the given selector.
+func (l ResourceIndexer[T]) List(selector labels.Selector) (ret []T, err error) {
+	// ListAllByNamespace reverts to ListAll on empty namespaces
```
I think it’s safe enough to rely on this.

skitt replied:
Done. I’ve filed https://github.com/kubernetes/kubernetes/issues/124412 to track the unit tests (as a separate PR).

Looking at `listers.go`, I see the structs and functions further down are tantalisingly similar to the generic versions I’m adding here... But I reckon addressing that is worth a separate PR.

thockin replied:
Thanks

skitt commented:
/retest-required

skitt commented on staging/src/k8s.io/code-generator/cmd/lister-gen/generators/lister.go:
```diff
@@ -345,35 +321,13 @@ type $.type|public$NamespaceLister interface {
 var namespaceListerStruct = `
 // $.type|private$NamespaceLister implements the $.type|public$NamespaceLister
 // interface.
 type $.type|private$NamespaceLister struct {
```
The code is identical but the purposes are different. The private structs are hooks for expansions, and if we merge the two then the functions from both leak into each other (in particular, the function in the non-namespaced lister that gives a namespaced lister)...

Of course I can’t find any actual usage of expansions on listers so perhaps we could just delete that mechanism wholesale.

thockin replied:
Not sure I follow.  Let's pick on Service as an example.

if I add a `FooBar` method to `ServiceListerExpansion`, then `NewServiceLister()` (which returns a generated wrapper of `listers.ResourceIndexer[*v1.Service]`) fails to compile because  `ResourceIndexer[*v1.Service]` does not have a `FooBar` method and no longer satisfies `ServiceLister`.

How would someone use the extensions?  I'm not seeing it.

skitt replied:
It works in a similar fashion to `client-gen` expansions; see `pod_expansion.go` for example.

For a Service lister, you’d add `service_expansion.go` containing

```go
package v1

// ServiceListerExpansion allows custom methods to be added to
// ServiceLister.
type ServiceListerExpansion interface {
	FooBar()
}

// ServiceNamespaceListerExpansion allows custom methods to be added to
// ServiceNamespaceLister.
type ServiceNamespaceListerExpansion interface {
}

func (s *serviceLister) FooBar() {
	// Play that funky music
}
```

Then run `lister-gen`, which will notice that there is an expansion defined for the Service lister, and skip it when generating `expansion_generated.go`. Then everything compiles, and `NewServiceLister()` returns an instance with a `FooBar` implementation alongside the other methods.

(The two expansion interfaces are handled together by `lister-gen`; if one is defined manually, the other needs to be too.)

skitt replied:
I’ve added an expansion example in the codegen examples.

skitt replied:
You’re welcome! I’d rather spend time discussing these things than have PRs approved without detailed review… Next up is #121439 ;-)

skitt committed:
Improve the lister function documentation

In particular, document that ListAllByNamespace delegates to ListAll
if no namespace is specified.

Signed-off-by: Stephen Kitt <skitt@redhat.com>

skitt committed:
Add a lister-gen expansion example

Signed-off-by: Stephen Kitt <skitt@redhat.com>

thockin reviewed (commented):
Thanks!

/lgtm
/approve
